File: golangR2/solution/solution0004.go
package solution

import "math"

func FindMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	//确保nums1比nums2短
	if len(nums1) > len(nums2) {
		tempNums := nums1
		nums1 = nums2
		nums2 = tempNums
	}

	m := len(nums1)
	n := len(nums2)

	totalLeft := m + (n-m+1)/2

	left, right := 0, m

	//在第一个数组nums1的区间[0,m]中查找恰当的分割线，满足nums1[i-1]<=nums2[j] && nums2[j-1]<=nums[i]
	//i-1表示在分割线左边
	for left < right {
		i := left + (right-left+1)/2
		j := totalLeft - i
		if nums1[i-1] > nums2[j] {
			right = i - 1
		} else {
			left = i
		}
	}

	i := left
	j := totalLeft - i
	var nums1LeftMax, nums1RightMin, nums2LeftMax, nums2RightMin int
	if i == 0 {
		nums1LeftMax = math.MinInt
	} else {
		nums1LeftMax = nums1[i-1]
	}

	if i == m {
		nums1RightMin = math.MaxInt
	} else {
		nums1RightMin = nums1[i]
	}

	if j == 0 {
		nums2LeftMax = math.MinInt
	} else {
		nums2LeftMax = nums2[j-1]
	}

	if j == n {
		nums2RightMin = math.MaxInt
	} else {
		nums2RightMin = nums2[j]
	}

	if (m+n)%2 == 1 {
		return math.Max(float64(nums1LeftMax), float64(nums2LeftMax))
	} else {
		return (math.Max(float64(nums1LeftMax), float64(nums2LeftMax)) + math.Min(float64(nums1RightMin), float64(nums2RightMin))) / 2
	}
}

File: JavaR3/src/solution/Solution0004.java
package solution;

public class Solution0004 {
	public double findMedianSortedArrays(int[] nums1, int[] nums2) {
		int m = nums1.length, n = nums2.length;

		int k1 = (m+n+1)/2;
		int k2 = (m + n + 2) / 2;

		//奇数和偶数转化为同一种情况
		return (getKth(0, m-1, 0,n-1, nums1, nums2, k1)
			+ getKth(0, m-1,0,n-1, nums1, nums2, k2)) * 0.5;

	}

	//寻找第k个数，每次删除k/2个数，直到k=1
	public static int getKth(int start1, int end1, int start2, int end2, int[] nums1, int[] nums2, int k) {
		int len1 = end1 - start1 + 1;
		int len2 = end2 - start2 + 1;
		if (len1 == 0 && len2 != 0) {
			return nums2[start2 + k - 1];
		}
		if (len2 == 0 && len1 != 0) {
			return nums1[start1 + k - 1];
		}

		while (k > 1) {

			int i = start1 + Math.min(len1, k / 2) - 1;
			int j = start2 + Math.min(len2, k / 2) - 1;

			if (nums1[i] < nums2[j]) {
				k = k - (i - start1 + 1);
				start1 = i + 1;
			} else {
				k = k - (j - start2 + 1);
				start2 = j + 1;
			}

			len1 = end1 - start1 + 1;
			len2 = end2 - start2 + 1;

			//如果其中一个数组没有数了
			if (len1 == 0 && len2 != 0) {
				return nums2[start2 + k - 1];
			}
			if (len2 == 0 && len1 != 0) {
				return nums1[start1 + k - 1];
			}
		}

		return Math.min(nums1[start1], nums2[start2]);
	}
}
